File: crates/taflow-core/src/indicators/ease_of_movement.rs
use crate::error::{TaError, TaResult};
use crate::stream::validate_period;
use std::collections::VecDeque;

/// Simple-moving-average of range-scaled midpoint movement per unit of volume.
#[derive(Debug, Clone)]
pub struct EaseOfMovement {
    period: usize,
    divisor: f64,
    previous_midpoint: Option<f64>,
    movements: VecDeque<f64>,
    sum: f64,
    count: usize,
    value: Option<f64>,
}

impl EaseOfMovement {
    /// Create an indicator with a period and conventional volume divisor.
    pub fn new(period: usize, divisor: f64) -> TaResult<Self> {
        validate_period(period)?;
        if !divisor.is_finite() || divisor <= 0.0 {
            return Err(TaError::InvalidParameter {
                name: "divisor",
                value: divisor.to_string(),
                reason: "must be finite and > 0",
            });
        }
        Ok(Self {
            period,
            divisor,
            previous_midpoint: None,
            movements: VecDeque::with_capacity(period),
            sum: 0.0,
            count: 0,
            value: None,
        })
    }

    /// Append one high/low/volume bar and return the averaged movement.
    pub fn append(&mut self, high: f64, low: f64, volume: f64) -> Option<f64> {
        self.count += 1;
        let midpoint = f64::midpoint(high, low);
        let Some(previous) = self.previous_midpoint.replace(midpoint) else {
            return None;
        };
        let movement = if volume == 0.0 {
            0.0
        } else {
            (midpoint - previous) * (high - low) * self.divisor / volume
        };
        if self.movements.len() == self.period {
            self.sum -= self.movements.pop_front().expect("full movement window");
        }
        self.movements.push_back(movement);
        self.sum += movement;
        self.value = (self.movements.len() == self.period).then(|| self.sum / self.period as f64);
        self.value
    }

    /// Return the latest averaged movement, or `None` during warm-up.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Return the number of processed bars.
    pub fn len(&self) -> usize {
        self.count
    }

    /// Return whether no bars have been processed.
    pub fn is_empty(&self) -> bool {
        self.count == 0
    }

    /// Clear midpoint, rolling mean, count, and latest-value state.
    pub fn reset(&mut self) {
        self.previous_midpoint = None;
        self.movements.clear();
        self.sum = 0.0;
        self.count = 0;
        self.value = None;
    }
}

File: python/taflow/indicators/ease_of_movement.py
"""Native-backed Ease of Movement adapter."""

from typing import Any

import numpy as np

from .._native import EaseOfMovementOperator as _Native
from .._series import as_float64_series


class EaseOfMovement:
    """Average range-scaled midpoint movement per unit of volume.

    ``high``, ``low``, and ``volume`` are required equal-length chronological
    series and may all be empty for a fresh stream. Rust owns midpoint change,
    rolling mean, warm-up, and aligned output. The raw movement is
    ``midpoint_change * range * divisor / volume`` and is averaged over
    ``period`` bars. Wickra ``EaseOfMovement`` is the oracle. ``compute`` returns one float
    array, ``value`` is the latest scalar or ``None`` during warm-up, and
    lifecycle mutators return ``self``.
    """

    def __init__(
        self,
        high: Any,
        low: Any,
        volume: Any,
        period: int = 14,
        divisor: float = 100_000_000.0,
    ) -> None:
        """Initialize aligned histories, averaging period, and output scale."""
        self._state = _Native(period, divisor)
        self.extend(high, low, volume)

    def append(self, high: float, low: float, volume: float) -> "EaseOfMovement":
        """Append one high/low/volume bar and return this adapter."""
        self._state.append(float(high), float(low), float(volume))
        return self

    def extend(self, high: Any, low: Any, volume: Any) -> "EaseOfMovement":
        """Append equal-length high, low, and volume histories."""
        arrays = tuple(as_float64_series(series) for series in (high, low, volume))
        if len({len(array) for array in arrays}) != 1:
            raise ValueError("high, low, and volume must have equal lengths")
        self._state.extend(*arrays)
        return self

    def compute(self) -> np.ndarray:
        """Return the aligned Ease of Movement history."""
        return self._state.compute()

    @property
    def value(self) -> float | None:
        """Return the latest movement value, or ``None`` during warm-up."""
        return self._state.value

    def reset(self) -> "EaseOfMovement":
        """Restore fresh native state and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of processed bars."""
        return len(self._state)
